### PR TITLE
Database

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: src/database/database.cpp [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <Poco/Data/SQLite/Connector.h>
+
+#include "database.hpp"
+
+using namespace Poco::Data::Keywords;
+
+namespace {
```
Comment (al1img) [inline src/database/database.cpp:14]: Could it be private member of Database?

File: src/database/database.cpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+ **********************************************************************************************************************/
+
+aos::Error Database::Init(const std::string& dbPath)
+{
+    if (mSession && mSession->isConnected()) {
+        return aos::ErrorEnum::eNone;
+    }
+
+    try {
+
```
Comment (al1img) [inline src/database/database.cpp:29]: Remove empty line

File: src/database/database.cpp [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        Poco::Data::SQLite::Connector::registerConnector();
+        mSession = std::optional<Poco::Data::Session>(Poco::Data::Session("SQLite", dbPath));
+        CreateTables();
+
```
Comment (al1img) [inline src/database/database.cpp:33]: ditto

File: src/database/database.cpp [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+aos::Error Database::AddCertInfo(const aos::String& certType, const aos::iam::certhandler::CertInfo& certInfo)
+{
+    if (!mSession || !mSession->isConnected()) {
+        return aos::ErrorEnum::eFailed;
+    }
+
+    try {
+
```
Comment (al1img) [inline src/database/database.cpp:48]: remove

File: src/database/database.cpp [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        *mSession
+            << "INSERT INTO certificates (type, issuer, serial, certURL, keyURL, notAfter) VALUES (?, ?, ?, ?, ?, ?);",
+            bind(ConvertToCertInfo(certType, certInfo)), now;
+
```
Comment (al1img) [inline src/database/database.cpp:52]: ditto

File: src/database/database.cpp [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    } catch (const Poco::Exception& e) {
+        return AOS_ERROR_WRAP(aos::ErrorEnum::eFailed);
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error Database::AddCertInfo(const aos::String& certType, const aos::iam::certhandler::CertInfo& certInfo)
+{
+    if (!mSession || !mSession->isConnected()) {
```
Comment (al1img) [inline src/database/database.cpp:43]: It is not necessary to check these conditions every time.

File: src/database/database.cpp [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                 "issuer BLOB NOT NULL,"
+                 "serial BLOB NOT NULL,"
+                 "certURL TEXT,"
+                 "keyURL TEXT,"
+                 "notAfter TIMESTAMP,"
+                 "PRIMARY KEY (issuer, serial));",
+        now;
+}
+
+Database::CertInfo Database::ConvertToCertInfo(
```
Comment (al1img) [inline src/database/database.cpp:175]: The names are ambiguous. Consider renaming conversion functions to something like: `ToAosCertInfo`, `FromAosCertInfo`, or something like that. Return `Database::CertInfo` through parameters.

File: tests/database/database_test.cpp [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+protected:
+    std::string mFileName = "test.db";
+    Database    mDB;
+};
+
+/***********************************************************************************************************************
+ * Tests
+ **********************************************************************************************************************/
+
+TEST_F(DatabaseTest, Init)
```
Comment (al1img) [inline tests/database/database_test.cpp:36]: Useless test. Init is tested in other test cases.

File: tests/database/database_test.cpp [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+{
+    EXPECT_EQ(mDB.Init(mFileName), aos::ErrorEnum::eNone);
+
+    EXPECT_EQ(mDB.Init(mFileName), aos::ErrorEnum::eNone);
+}
+
+TEST_F(DatabaseTest, AddCertInfo)
+{
+    aos::iam::certhandler::CertInfo certInfo;
```
Comment (al1img) [inline tests/database/database_test.cpp:45]: Empty line after var definition.

File: tests/database/database_test.cpp [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    EXPECT_EQ(mDB.AddCertInfo("type", certInfo), aos::ErrorEnum::eNone);
+}
+
+TEST_F(DatabaseTest, RemoveCertInfo)
+{
+    EXPECT_EQ(mDB.RemoveCertInfo("type", "certURL"), aos::ErrorEnum::eFailed);
+
+    EXPECT_EQ(mDB.Init(mFileName), aos::ErrorEnum::eNone);
+
+    aos::iam::certhandler::CertInfo certInfo;
```
Comment (al1img) [inline tests/database/database_test.cpp:73]: ditto

File: tests/database/database_test.cpp [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+    EXPECT_EQ(mDB.RemoveCertInfo("type", "certURL"), aos::ErrorEnum::eNone);
+}
+
+TEST_F(DatabaseTest, RemoveAllCertsInfo)
+{
+    EXPECT_EQ(mDB.RemoveAllCertsInfo("type"), aos::ErrorEnum::eFailed);
+
+    EXPECT_EQ(mDB.Init(mFileName), aos::ErrorEnum::eNone);
+
+    aos::iam::certhandler::CertInfo certInfo;
```
Comment (al1img) [inline tests/database/database_test.cpp:91]: ditto

File: tests/database/database_test.cpp [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    EXPECT_EQ(mDB.GetCertInfo(certInfo.mIssuer, certInfo.mSerial, certInfo), aos::ErrorEnum::eNotFound);
+
+    certInfo.mIssuer   = StringToDN("issuer");
+    certInfo.mSerial   = StringToDN("serial");
+    certInfo.mCertURL  = "certURL";
+    certInfo.mKeyURL   = "keyURL";
+    certInfo.mNotAfter = aos::Time::Now();
+    EXPECT_EQ(mDB.AddCertInfo("type", certInfo), aos::ErrorEnum::eNone);
+
+    aos::iam::certhandler::CertInfo certInfo2;
```
Comment (al1img) [inline tests/database/database_test.cpp:127]: ditto

File: tests/database/database_test.cpp [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+    EXPECT_EQ(mDB.Init(mFileName), aos::ErrorEnum::eNone);
+
+    EXPECT_EQ(mDB.GetCertInfo(certInfo.mIssuer, certInfo.mSerial, certInfo), aos::ErrorEnum::eNotFound);
+
+    certInfo.mIssuer   = StringToDN("issuer");
+    certInfo.mSerial   = StringToDN("serial");
+    certInfo.mCertURL  = "certURL";
+    certInfo.mKeyURL   = "keyURL";
+    certInfo.mNotAfter = aos::Time::Now();
+    EXPECT_EQ(mDB.AddCertInfo("type", certInfo), aos::ErrorEnum::eNone);
```
Comment (al1img) [inline tests/database/database_test.cpp:125]: Add empty line

File: tests/database/database_test.cpp [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    aos::iam::certhandler::CertInfo certInfo2;
+    certInfo2.mIssuer   = StringToDN("issuer2");
+    certInfo2.mSerial   = StringToDN("serial2");
+    certInfo2.mCertURL  = "certURL2";
+    certInfo2.mKeyURL   = "keyURL2";
+    certInfo2.mNotAfter = aos::Time::Now();
+
+    EXPECT_EQ(mDB.AddCertInfo("type", certInfo2), aos::ErrorEnum::eNone);
+
+    aos::iam::certhandler::CertInfo certInfoStored {};
```
Comment (al1img) [inline tests/database/database_test.cpp:136]: empty line after var definiton.

File: src/database/database.cpp (hunk 1 above)
Comment (MykolaSolyanko) [reply src/database/database.cpp:14]: Done

File: src/database/database.cpp (hunk 3 above)
Comment (MykolaSolyanko) [reply src/database/database.cpp:33]: Done

File: src/database/database.cpp (hunk 4 above)
Comment (MykolaSolyanko) [reply src/database/database.cpp:48]: Done

File: src/database/database.cpp (hunk 5 above)
Comment (MykolaSolyanko) [reply src/database/database.cpp:52]: Done

File: src/database/database.cpp (hunk 6 above)
Comment (MykolaSolyanko) [reply src/database/database.cpp:43]: Remove

File: src/database/database.cpp (hunk 7 above)
Comment (MykolaSolyanko) [reply src/database/database.cpp:175]: Done

File: tests/database/database_test.cpp (hunk 8 above)
Comment (MykolaSolyanko) [reply tests/database/database_test.cpp:36]: Remove

File: tests/database/database_test.cpp (hunk 9 above)
Comment (MykolaSolyanko) [reply tests/database/database_test.cpp:45]: Done

File: tests/database/database_test.cpp (hunk 10 above)
Comment (MykolaSolyanko) [reply tests/database/database_test.cpp:73]: Done

File: tests/database/database_test.cpp (hunk 11 above)
Comment (MykolaSolyanko) [reply tests/database/database_test.cpp:91]: Done

File: tests/database/database_test.cpp (hunk 12 above)
Comment (MykolaSolyanko) [reply tests/database/database_test.cpp:127]: Done

File: tests/database/database_test.cpp (hunk 13 above)
Comment (MykolaSolyanko) [reply tests/database/database_test.cpp:125]: Done

File: tests/database/database_test.cpp (hunk 14 above)
Comment (MykolaSolyanko) [reply tests/database/database_test.cpp:125]: Done

File: src/CMakeLists.txt [hunk 15]
```diff
@@ -23,6 +23,7 @@ include_directories(${CMAKE_CURRENT_SOURCE_DIR})
 
 add_subdirectory(app)
 add_subdirectory(logger)
+add_subdirectory(database)
```
Comment (mlohvynenko) [inline src/CMakeLists.txt:26]: sort by alphabet

File: src/CMakeLists.txt [hunk 16]
```diff
@@ -40,7 +41,7 @@ add_executable(${TARGET} ${SOURCES})
 # Libraries
 # ######################################################################################################################
 
-target_link_libraries(${TARGET} PUBLIC app logger)
+target_link_libraries(${TARGET} PUBLIC app logger database)
```
Comment (mlohvynenko) [inline src/CMakeLists.txt:44]: sort libs by alphabet

File: src/database/database.cpp [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+{
+    if (mSession && mSession->isConnected()) {
+        return aos::ErrorEnum::eNone;
+    }
+
+    try {
+        Poco::Data::SQLite::Connector::registerConnector();
+        mSession = std::optional<Poco::Data::Session>(Poco::Data::Session("SQLite", dbPath));
+        CreateTables();
+    } catch (const Poco::Exception& e) {
```
Comment (mlohvynenko) [inline src/database/database.cpp:28]: It's worth logging error message

File: src/database/database.cpp [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+        return AOS_ERROR_WRAP(aos::ErrorEnum::eFailed);
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error Database::AddCertInfo(const aos::String& certType, const aos::iam::certhandler::CertInfo& certInfo)
+{
+    try {
+        *mSession
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply src/database/database.cpp:38]: this SQLite, SQLite driver itself supports blocking, so not necessary to use blocking mechanism

File: src/CMakeLists.txt (hunk 16 above)
Comment (MykolaSolyanko) [reply src/CMakeLists.txt:44]: Done

File: src/CMakeLists.txt (hunk 15 above)
Comment (MykolaSolyanko) [reply src/CMakeLists.txt:26]: Done

**Review by mlohvynenko — APPROVED** [review]

**Pull request merged**